src/app/src/main/kotlin/org/vaccineimpact/api/app/serialization/DataTableDeserializer.kt

```diff
@@ -0,0 +1,157 @@
+            val value = Deserializer().deserialize(raw.trim(), targetType)
+            value
+        }
+        catch (e: Exception)
+        {
+            problems.add(ErrorInfo(
+                    "csv-bad-data-type:$row:$column",
+                    "Unable to parse '${raw.trim()}' as ${targetType.javaClass.simpleName} (Row $row, column $column)"
+            ))
+            null
+        }
+    }
+
+    private fun checkHeaders(actualHeaders: List<String>): List<HeaderDefinition>
+    {
+        val problems = mutableListOf<ErrorInfo>()
+        var index = 0
+        while (index < maxOf(headerCount, actualHeaders.size))
+        {
+            val expected = headers.getOrNull(index)
+            val actual = actualHeaders.getOrNull(index)?.trim()
+
+            if (expected != null)
+            {
+                val expectedName = expected.name
+                if (actual != null)
+                {
+                    if (expectedName != actual)
+                    {
+                        problems.add(ErrorInfo("csv-unexpected-header", "Expected column header '$expectedName'; found '$actual' instead (column $index)"))
+                    }
+                }
+                else
+                {
+                    problems.add(ErrorInfo("csv-missing-header", "Not enough column headers were provided. Expected a '$expectedName' header."))
+                }
+            }
+            else if (actual != null && !extraHeadersAllowed)
+            {
+                problems.add(ErrorInfo("csv-unexpected-header", "Too many column headers were provided. Unexpected '$actual' header."))
+            }
+            index += 1
+        }
+
+        if (problems.any())
+        {
+            throw ValidationError(problems)
+        }
+        return headers
+    }
+
+    protected open fun getActualHeaderDefinitions(actualHeaders: List<String>): List<HeaderDefinition>
+    {
+        return headers
+    }
+
+    companion object
+    {
+        fun <T : Any> deserialize(
+                body: String,
+                type: KClass<T>,
+                serializer: Serializer = Serializer.instance
+        ): Sequence<T>
+        {
+            return getDeserializer(type, serializer).deserialize(body)
+        }
+
+        private fun <T: Any> getDeserializer(type: KClass<T>, serializer: Serializer): DataTableDeserializer<T>
+        {
+            val constructor = type.primaryConstructor
+                    ?: throw Exception("Cannot deserialize to type ${type.simpleName} - it has no primary constructor")
+            val headers = constructor.parameters
+                    .map { HeaderDefinition(serializer.convertFieldName(it.name!!), it.type) }
+
+            return if (type.findAnnotation<FlexibleColumns>() != null)
+            {
+                // If the last argument is a map, get the type of the values the map stores
+                val flexibleType = constructor.parameters.last().type.arguments.last().type!!
```

Comment:
If a type is marked with the 'flexible' annotation but doesn't have a map as the last property type this call to `last()` will throw an error. Might be worth checking for this and throwing an informative error in this case, e.g.

      val args = constructor.parameters.last().type.arguments
      if (!args.any)
      {
        throw some informative error here
      }

Reply:
There are lots of if statements needed to check this properly, and some of them I couldn't even figure out how to write, hence why I went with an annotation based approach. So I've just wrapped the whole thing in a `try...catch` and made it throw a more helpful exception if anything goes wrong.